### PR TITLE
Read version 51 (JDK 7) class files.

### PR DESCRIPTION
**Pull request by JamesIry**
This commit makes the ClassFileReader/ICodeReader parse class files
from JDK 7 (class file version 51). It does that by skipping over
the method handle related entries in the constant pool and by doing
some dummy processing on invoke dynamic instructions. The inliner
is updated to not try to inline a method with an invoke dynamic
instruction. A place holder INVOKE_DYNAMIC instruction is added to ICode
but it is designed to create an error if there's ever any attempt to
analyze it. Because the inliner is the only phase that ever tries
to analyze ICode instructions not generated from Scala source and
because Scala source will never emit an INVOKE_DYNAMIC, the place
holder INVOKE_DYNAMIC should never cause any errors.

A test is included that generates a class file with INVOKE_DYNAMIC
and then compiles Scala code that depends on it.

### CODE REVIEW

**Comment by JamesIry** [comment]
subsumes https://github.com/scala/scala/pull/2243, putting it into the 2.10.x branch

**Comment by JamesIry** [comment]
review @adriaanm

**Comment by paulp** [comment]
LGTM

File: src/compiler/scala/tools/nsc/symtab/classfile/ICodeReader.scala [hunk 1]
```diff
@@ -506,6 +506,12 @@ abstract class ICodeReader extends ClassfileParser {
             code.emit(UNBOX(toTypeKind(m.info.resultType)))
           else
             code.emit(CALL_METHOD(m, Static(false)))
+        case JVM.invokedynamic  =>
+          // TODO, this is just a place holder. A real implementation must parse the class constant entry
+          containsInvokeDynamic = true
+          val poolEntry = in.nextChar
+          in.skip(2)
+          code.emit(INVOKE_DYNAMIC(poolEntry))
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/symtab/classfile/ICodeReader.scala:515]: Can we get more in the habit of logging something when "placeholder" code paths are triggered?

**Comment by paulp** [comment]
I'm pulling back my LGTM holding out for a call to log() where indicated, then it will LGTM again.

**Comment by JamesIry** [comment]
Added a debuglog

**Pull request merged**